pkg/types/dsse/v0.0.1/entry.go

```diff
@@ -0,0 +1,420 @@
+			}
+		}
+		// Not all in-toto statements will contain a SLSA provenance predicate.
+		// See https://github.com/in-toto/attestation/blob/main/spec/README.md#predicate
+		// for other predicates.
+		if predicate, err := parseSlsaPredicate(decodedPayload); err == nil {
+			if predicate.Predicate.Materials != nil {
+				for _, s := range predicate.Predicate.Materials {
+					for alg, ds := range s.Digest {
+						result = append(result, alg+":"+ds)
+					}
+				}
+			}
+		}
+	default:
+		log.Logger.Infof("Unknown DSSE envelope payloadType: %s", v.env.PayloadType)
+	}
+	return result, nil
+}
+
+func parseStatement(p []byte) (*in_toto.Statement, error) {
+	ps := in_toto.Statement{}
+	if err := json.Unmarshal(p, &ps); err != nil {
+		return nil, err
+	}
+	return &ps, nil
+}
+
+func parseSlsaPredicate(p []byte) (*in_toto.ProvenanceStatement, error) {
+	predicate := in_toto.ProvenanceStatement{}
+	if err := json.Unmarshal(p, &predicate); err != nil {
+		return nil, err
+	}
+	return &predicate, nil
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	it, ok := pe.(*models.DSSE)
+	if !ok {
+		return errors.New("cannot unmarshal non DSSE v0.0.1 type")
+	}
+
+	dsseObj := &models.DSSEV001Schema{}
+
+	if err := types.DecodeEntry(it.Spec, dsseObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := dsseObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	// either we have just proposed content or the canonicalized fields
+	if dsseObj.ProposedContent == nil {
+		// then we need canonicalized fields, and all must be present (if present, they would have been validated in the above call to Validate())
+		if dsseObj.EnvelopeHash == nil || dsseObj.PayloadHash == nil || len(dsseObj.Signatures) == 0 {
+			return errors.New("either proposedContent or envelopeHash, payloadHash, and signatures must be present")
+		}
+		v.DSSEObj = *dsseObj
+		return nil
+	}
+	// if we're here, then we're trying to propose a new entry so we check to ensure client's aren't setting server-side computed fields
+	if dsseObj.EnvelopeHash != nil || dsseObj.PayloadHash != nil || len(dsseObj.Signatures) != 0 {
+		return errors.New("either proposedContent or envelopeHash, payloadHash, and signatures must be present but not both")
+	}
+
+	env := &dsse.Envelope{}
+	if err := json.Unmarshal([]byte(*dsseObj.ProposedContent.Envelope), env); err != nil {
+		return err
+	}
+
+	if len(env.Signatures) == 0 {
+		return errors.New("DSSE envelope must contain 1 or more signatures")
+	}
+
+	allPubKeyBytes := make([][]byte, 0)
+	for _, publicKey := range dsseObj.ProposedContent.PublicKeys {
+		allPubKeyBytes = append(allPubKeyBytes, publicKey)
+	}
+
+	sigToKeyMap, err := verifyEnvelope(allPubKeyBytes, env)
+	if err != nil {
+		return err
+	}
+
+	// we need to ensure we canonicalize the ordering of signatures
+	sortedSigs := make([]string, 0, len(sigToKeyMap))
+	for sig := range sigToKeyMap {
+		sortedSigs = append(sortedSigs, sig)
+	}
+	sort.Strings(sortedSigs)
```

Comment:
To confirm, these sigs are base64 encoded already, correct? The DSSE spec says signatures must be base64 encoded, I assume the DSSE verifier checks this?

Reply:
`verifyEnvelope` will do all of that and would fail above on line 210 if they can't be used to verify the signatures.